pkg/scheduler/factory/factory.go

```diff
@@ -186,6 +191,16 @@ func NewConfigFactory(
 		disablePreemption:              disablePreemption,
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.DynamicVolumeThreshold) {
+		volumePlugins := volumeplugins.ProbeVolumePlugins()
+		c.volumePluginMgr = &volume.VolumePluginMgr{}
```

Comment:
Why do we need to add yet another volume plugin to the guts of the scheduler? I don't remember this was discussed in the design doc.

Reply:
So this is not just another volume plugin. This brings in volume plugin manager, so as we can identify which volume type is being used in pod without hardcoding each volume plugin - https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithm/predicates/predicates.go#L493 . This also lets us identify which resource key will be present in node allocatable to check volume attach limit etc. 

I am going to update the design to reflect this. This was brought up during proposal discussions and I thought I addressed it...

Reply:
I have explicitly called this out in design doc - https://github.com/gnufied/community/blob/055f8da12bbebca787445da4fcd7352996278712/contributors/design-proposals/storage/dynamic_volume_limit.md#changes-to-scheduler now

To some extent, I do not understand opposition to adding volume plugin manager in scheduler because alternative one way or another is to hardcode and switch..case based on volume type.

There are several ways we can workaround this, if it is a blocker:

1. We can hardcode for in-tree volume types in the predicate (Azure, AWS, GCE-PD and CSI). CSI will require special handling because `VolumeLimitKey` has to be generated by in-tree CSI plugin that exists in k8s.  There are two ways to fix CSI handling:
    - Rather than importing volume plugin manager in scheduler, we can just import CSI in-tree volume plugin. So we can hardcode type checks for aws, gce, azure and for CSI we can call `VolumeLimitKey` function defined inside CSI plugin to get the key. Pros of this approach is: we can reuse `VolumeLimitKey` logic between scheduler, kubelet and other components. Also we don't have to import volume plugin manager, because that imports all volume plugins.
    - Another option is - we can duplicate code inside `VolumeLimitKey` for CSI and copy it into scheduler. Problem with this approach is - overtime the `VolumeLimitKey` implemntation in volume plugin and scheduler may drift apart. :(

2. Another option @liggitt has proposed is https://github.com/kubernetes/community/pull/2051#discussion_r190778327 . What that does is - it adds a `ResourceRequirement` field to `CSIPersistentVolumeSource` with `{"storage-attach-request-aws-ebs": 1 } `.  There are couple of problems with this I think: 
    - Who decides key name inside `ResourceRequirement` map of `CSIPersistentVolumeSource`? For various reasons - it can't be same as CSI plugin name and plugin name is only string that uniquely identifies a CSI driver.  Also, CSI provisioner and plugin on node are two components, so we are still left with duplicating logic that will generate `VolumeLimitKey` in multiple places.
    - I can't help but think that only reason `CSIPersistentVolumeSource` carries `ResourceRequirement` is to carry key that will match resource name in node's allocatable for CSI.  So it is not a resource requirement of PV in typical sense.

Reply:
@msau42 proposed another option that is, we hardcode volume types for AWS, GCE-PD, Azure and CSI. But for CSI we can write a util function that is used both in scheduler and CSI volume plugin to generate `VolumeLimitKey` . cc @liggitt do you think this works?

Reply:
>The provisioner can be told what resource name to use in a storageclass parameter (as it does for fstype) or as a startup option (as it does with --volume-name-prefix) and set it on the PV when it constructs it here:

The resource name has to be available in volume plugin driver running on the node too(so as it can be registered in node's allocatable).  The problem is - a CSI plugin is CO agnostic component and it doesn't know anything about Kubernetes or resource name restriction that applies to names chosen by it.

> as those transition to CSI, I expected the CSI replacements for AWS/GCE/Azure to be able to continue indicating the same attach count resource name for the CSI PVs they provision. That implies the resource name is specified in the CSI PV, not derived from the CSI plugin or driver name

Yeah I definitely see value in this. However current in-tree to CSI migration plan is considering having CSI spec of a in-tree PV as a annotation in the PV itself. From scheduler perspective - this will allow us to consider a non-CSI volume as CSI and do the counting.

Reply:
Remove volume plugin manager from scheduler